index.js

```diff
@@ -143,44 +139,75 @@ ParallelHook.prototype[$map] = function(f){
   return ParallelHook(map(f)(this.hook));
 }
 
-/* c8 ignore next */
-const crash = e => Future(() => { throw e });
-
-ParallelHook.prototype[$ap] = function(mf){
-  return ParallelHook(Hook(c => {
-    let consume = noop;
-    const rf = mf.hook.run(f => consume !== noop ? consume (f) : (
-      Future((rej, res) => {
-        consume = x => map(y => (res(y), y))(c(f(x)));
-        return noop;
-      })
+const CustomFuture = function (interpret) {
+  this._interpret = interpret;
+}
+
+CustomFuture.prototype = Object.create (Future.prototype);
+
+ParallelHook.prototype[$ap] = function(parallelFunctionHook){
+  const withValue = runHook (sequential (this));
+  const withFunction = runHook (sequential (parallelFunctionHook));
+  return ParallelHook(Hook(consumeResult => new CustomFuture ((rec, rej, res) => {
+    let deferred = null, cont = null, cancelled = false, cancelValue = null, cancelFunction = null;
+
+    const defer = value => {
+      deferred = cache (never);
+      deferred.value = value;
+      return deferred;
+    };
+
+    const consume = resource => new CustomFuture ((rec, rej, res) => (
+      consumeResult (resource) ._interpret ( x => {rec (x); deferred.crash (x)}
+                                           , x => {rej (x); deferred.reject (x)}
+                                           , x => {res (x); deferred.resolve (x)} )
     ));
-    const rx = this.hook.run(x => consume !== noop ? consume (x) : (
-      Future((rej, res) => {
-        consume = f => map(y => (res(y), y))(c(f(x)));
-        return noop;
-      })
+
+    const eventuallyConsumeValue = withValue (val => (
+      deferred ? consume (deferred.value (val)) : defer (val)
     ));
-    const transformation = {
-      cancel: noop,
-      context: rx.context,
-      /* c8 ignore next */
-      rejected: () => never,
-      resolved: () => never,
-      /* c8 ignore next */
-      toString: () => `parallelHookTransform(${rx.toString()})`,
-      run: early => {
-        const action = Object.create(transformation);
-        action.cancel = rx._interpret(
-          e => early(crash(e), action),
-          x => early(reject(x), action),
-          x => early(resolve(x), action),
-        );
-        return action;
-      },
+
+    const eventuallyConsumeFunction = withFunction (func => (
+      deferred ? consume (func (deferred.value)) : defer (func)
+    ));
+
+    /* c8 ignore next 3 */
+    const crash = x => {
+      cancel ();
+      rec (x);
+    };
```

Comment:
I did not write tests for the crash code paths, because writing assertions for crashed Futures is difficult, and the code path is relatively simple.